reStream.sh

```diff
@@ -86,6 +80,60 @@ if ! ssh_cmd true; then
     exit 1
 fi
 
+rm_version="$(ssh_cmd cat /sys/devices/soc0/machine)"
+
+case "$rm_version" in
+    "reMarkable 1.0")
+        width=1408
+        height=1872
+        bytes_per_pixel=2
+        pixel_format="rgb565le"
+        # calculate how much bytes the window is
+        window_bytes="$((width * height * bytes_per_pixel))"
+        # read the first $window_bytes of the framebuffer
+        head_fb0="dd if=/dev/fb0 count=1 bs=$window_bytes 2>/dev/null"
+        ;;
+    "reMarkable 2.0")
+        pixel_format="gray8"
+        width=1872
+        height=1404
+        bytes_per_pixel=1
+
+        # calculate how much bytes the window is
+        window_bytes="$((width * height * bytes_per_pixel))"
+
+        # find xochitl's process
+        pid="$(ssh_cmd pidof xochitl)"
+        echo "xochitl's PID: $pid"
+
+        # find framebuffer location in memory
+        # it is actually the map allocated _after_ the fb0 mmap
+        read_address="grep -C1 '/dev/fb0' /proc/$pid/maps | tail -n1 | sed 's/-.*$//'"
+        skip_bytes_hex="$(ssh_cmd "$read_address")"
+        skip_bytes="$((0x$skip_bytes_hex + 8))"
+        echo "framebuffer is at 0x$skip_bytes_hex"
+
+        # carve the framebuffer out of the process memory
+        page_size=4096
+        window_start_blocks="$((skip_bytes / page_size))"
+        window_offset="$((skip_bytes % page_size))"
+        window_length_blocks="$((window_bytes / page_size + 1))"
+
+        # Using dd with bs=1 is too slow, so we first carve out the pages our desired
+        # bytes are located in, and then we trim the resulting data with what we need.
+        head_fb0="dd if=/proc/$pid/mem bs=$page_size skip=$window_start_blocks count=$window_length_blocks 2>/dev/null | tail -c+$window_offset | head -c $window_bytes"
```

Comment:
cool I just grabbed the latest stable busybox version from https://www.busybox.net/downloads/binaries/1.31.0-defconfig-multiarch-musl/busybox-armv7l  and replaced/linked to /bin/busybox.nosuid and it works!

Reply:
I would suggest testing with the version of busybox (or better yet coreutils-head) in entware since it'll make more sense to keep within the established ecosystem: https://github.com/Evidlo/remarkable_entware

Reply:
"opkg install head" was the first (dumb?) thing I tried. but even not considering the name conflict with built-in busybox, it returns unknown package anyways... waiting for someone more familiar with bash/shell to chime in, maybe a proper dd command is enough? (still don't understand the difference between head -c $bytes vs dd bs=$bytes count=1 ...)

Reply:
I kept the stock busybox and just installed debian's `head` from coreutils package ([armhf]( https://packages.debian.org/buster/coreutils))

Reply:
While it can be scripted to use debian's coreutils, the existing ecosystem on the remarkable (https://github.com/toltec-dev/toltec) uses entware, so it would make more sense to continue using it.

Reply:
I can confirm that it works with coreutils-head from entware.

Reply:
To get Portrait and Landscape working 100% after installing debian's head(), I had to do additional video settings:
```
# Rotate 90˚ clockwise if landscape=true
# Rotate 90˚ counter-clockwise and flip horiz+vert if landscape=false
if $landscape; then
    video_filters="$video_filters,transpose=2"
else
    video_filters="$video_filters,transpose=1,vflip,hflip"
fi
```

Reply:
> `head -c` not supported by `head` from busybox, requires armhf compiled binary, and can be replace by e.g. `/usr/local/bin/head`

About the head `head -c`, I would suggest to replace `head -c $window_bytes"` with `dd count=$window_bytes bs=1 2>/dev/null` which I tested and works on my up to date reMarkable2 without any additional binaries copied.

Reply:
dd with bs=1 and non-zero count is very slow.
OTOH you can just call dd twice, I use the following snippet here to do everything in a single ssh invocation (not for reStream, but you can use the dd commands):
```
width=1872
height=1404
bytes_per_pixel=1
window_bytes="$((width * height * bytes_per_pixel))"

echo "
pid=\$(pidof xochitl)
offset=\$(awk -F- '/\/dev\/fb0/ { getline; print \"0x\" \$1; } ' < /proc/\$pid/maps)
{ dd bs=1 skip=\$((offset)) count=0; dd bs=$window_bytes count=1; } < /proc/\$pid/mem 2>/dev/null
" | ssh $REMARKABLE
```